### ISSUE TITLE
Citrus v4.0.0

### ISSUE THREAD

**Issue by christophd** [citrusframework/citrus#922]
We should start thinking and discussing about a new major version of Citrus 4.0.0.

I suppose to migrate main branch to 4.0.0-SNAPSHOT and start working on that major version soon. At the same time update branch 3.x and continue with Citrus v3 there

Checklist for Citrus 4.0.0:

- [x] Use 4.0.0-SNAPSHOT on main branch
- [x] Use Java 17 as minimum requirement
- [x] Remove deprecated code
- [x] Remove vintage modules (e.g. citrus-arquillian, citrus-java-dsl)
- [x] Update to Spring 6.0 (see #907)
- [x] Update to Camel 4.0
- [x] Migrate from `javax.*` to `jakarta.*` (JEE10)
- [x] Move groupId from `com.consol.citrus` to `org.citrusframework`
- [ ] Update/Rewrite documentation
- [x] Update Jetty to v11 (see #871)
- [x] Move citrus-remote modules to separate repository
- [x] Move citrus-jdbc module to [citrus-db](https://github.com/citrusframework/citrus-db) repository

Optional:

- [ ] Remove support for JUnit4
- [ ] Update citrus-kubernetes module to use latest Kubernetes client
- [ ] Update citrus-docker module
- [ ] Remove citrus-restdocs module
- [ ] Move from TestNG to JUnit Jupiter (internal refactoring)

**Comment by christophd** [comment]
Things to discuss:

- Remove support for JUnit4
- Discontinue citrus-restdocs module
- Discontinue citrus-remote module
- Discontinue citrus-jmx module
- Discontinue citrus-rmi module
- Update citrus-kubernetes module to use latest Kubernetes client
- Update citrus-docker module

**Comment by bbortt** [comment]
I'm in 😉 #907 will break things pretty hard. and I can certainly work on this stuff, but only part-time. so, any help is much appreciated.

**Comment by bbortt** [comment]
also #871.

**Comment by turing85** [comment]
> Things to discuss:
> 
> ...
> * Discontinue citrus-remote module
> ...
> * Update citrus-docker module

What is the rationale for this, and what would be the alternative (especially when using YAKS is not an option?

**Comment by bbortt** [comment]
#904 will solve the following:

> - [ ] Use Java 17 as minimum requirement
> - [ ] Update to Spring 6.0 (see https://github.com/citrusframework/citrus/issues/907)

**Comment by christophd** [comment]
@turing85 citrus-remote is based on Apache Spark library which is not ready for jakarta.servlet API. We would need to switch to another lightweight web framework and basically rewrite the module. How about moving the citrus-remote modules to a separate GitHub repository?

This way we can continue with Citrus 4.0 and update the citrus-remote modules separately.

**Comment by bbortt** [comment]
@christophd maybe also:
- [ ] move testng tests to junit (internal refactoring)

**Comment by xhandalax** [comment]
Wird auch auf org.apache.httpcomponents.client5.httpclient5 aktualisiert?

**Comment by bbortt** [comment]
> Wird auch auf org.apache.httpcomponents.client5.httpclient5 aktualisiert?

@xhandalax that has already been done in #932.

**Comment by christophd** [comment]
A 1st milestone of Citrus 4.0.0 just has been released: v4.0.0-M1 

Please be aware that the Maven groupId has changed from `com.consol.citrus` to `org.citrusframework`

Now it is time to give it a try and report any issues. Many thanks to all contributors!

**Comment by bbortt** [comment]
I am working on this:
> Remove vintage modules (e.g. citrus-arquillian, citrus-java-dsl)

**Comment by bbortt** [comment]
@christophd I can confirm that the current commit on `master` (https://github.com/citrusframework/citrus/commit/2b8405ea6931e064b930964de0abaff012d00e47) builds the `citrus-simulator` upon `citrus-simulator-sample-combined`, as it's expected to do.

**Comment by christophd** [comment]
@bbortt v4.0.0-M2 just got released. Please let me know in case something is still missing for 4.0.0

I am working on updating and improving the documentation but this may be an ongoing task for the near future. So from my side we are good to cut a 4.0.0 release as soon as you folks are ready.

**Comment by bbortt** [comment]
> @bbortt v4.0.0-M2 just got released. Please let me know in case something is still missing for 4.0.0

already had do open another PR. sorry. we'll know as soon as the simulator builds. 

> I am working on updating and improving the documentation but this may be an ongoing task for the near future. So from my side we are good to cut a 4.0.0 release as soon as you folks are ready.

I would love to have a simulator milestone release beforehand as well. that would allow me to run our company-internal tests (we have a lot of tests based on citrus). I am sure we will find something (if not - even better). and then cut the release.

if that is ok for you folks to have some patience? 🙃

**Comment by apupier** [comment]
@bbortt Is the simulator in main branch now?

**Comment by bbortt** [comment]
> @bbortt Is the simulator in main branch now?

ahm yes. the `main` branch is fully JDK17 and citrusframework `4.0.0-M2` compatible.

there are some additional changes to the persistency layer contained in #177. and the REST API must be fully refactored (only partially done). but the "simulation core" is done, just the landscape around it might change a little.

**Comment by christophd** [comment]
Folks, I think we are finally good to release Citrus 4.0.0!

There is a last pending PR #1037 

Once this one is merged I will proceed with the release process

**Comment by christophd** [comment]
some issues related to running Citrus on Windows OS popped up: https://github.com/citrusframework/citrus/issues/1038

**Comment by bbortt** [comment]
https://github.com/citrusframework/citrus/releases/tag/v4.0.0

**Comment by christophd** [comment]
Yes, we made it! 🥳 🎉 Thank you everyone that helped to get this done! Citrus 4.0 is a major community achievement!

**Comment by ZuikovAV** [comment]
@christophd 
1. Could you tell me please, how to work now with before-suite and before-test actions  (TestRunnerBeforeSuiteSupport, TestRunnerBeforeTestSupport as i see were removed with citrus-java-dsl)?
2. Does any similar functionality like citrus.channel present at 4.0.0? MessageSelectionQueueChannel now is not working. I used it to set polling intervals for servers and now don't know what to do without this opportunity.
